File: sd/slcanx-cpp/examples/05_simple_fd.cpp
#include "slcanx.hpp"
#include <iostream>
#include <thread>
#include <chrono>

using namespace slcanx;

int main(int argc, char** argv) {
    std::string port = "COM3";
    if (argc > 1) port = argv[1];

    Slcanx slcan(port);

    slcan.close_channel(0);
    
    // Arb 500K, Data 2M
    slcan.set_bitrate(0, 500000);
    slcan.set_data_bitrate(0, 2000000);
    slcan.set_sample_point(0, 80.0, 80.0);
    
    slcan.open_channel(0);

    slcan.set_rx_callback([](uint8_t ch, const CanFrame& frame) {
        if (frame.fd) {
            std::cout << "Rx FD: ID=" << std::hex << frame.id << " Len=" << frame.data.size() << std::dec << std::endl;
        }
    });

    std::vector<uint8_t> data(64, 0xAA);
    CanFrame frame = CanFrame::new_fd(0x123, data, true); // BRS
    
    std::cout << "Sending FD frame..." << std::endl;
    slcan.send(0, frame);

    while(true) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    return 0;
}

File: sd/slcanx-rs/examples/05_simple_fd.rs
use clap::Parser;
use slcanx::{Slcanx, CanFrame};
use std::time::Duration;
use anyhow::Result;

#[derive(Parser)]
struct Args {
    #[arg(short, long, default_value = "COM3")]
    port: String,
    #[arg(short, long, default_value_t = 0)]
    channel: u8,
}

fn main() -> Result<()> {
    env_logger::init();
    let args = Args::parse();

    let slcan = Slcanx::new(&args.port, 115200, Duration::from_micros(125))?;
    let channel = slcan.get_channel(args.channel);

    channel.close()?;

    // Arb 500K, Data 2M
    channel.set_bitrate(500000)?;
    channel.set_data_bitrate(2000000)?;
    channel.set_sample_point(Some(80.0), Some(80.0))?;
    
    channel.open()?;

    // Send FD frame
    let data = vec![0xAA; 64];
    let frame = CanFrame::new_fd(0x123, &data, true); // BRS enabled
    println!("Sending FD frame...");
    channel.send(frame)?;

    println!("Listening...");
    loop {
        if let Some((ch, msg)) = slcan.recv_timeout(Duration::from_secs(1))? {
            if ch == args.channel {
                println!("Received: {:?}", msg);
            }
        }
    }
}
